backends/magma/ceed-magma.c

```diff
@@ -0,0 +1,806 @@
+      break;
+    case CEED_OWN_POINTER:
+      impl->dindices = (CeedInt *)indices;
+      ierr = magma_malloc_pinned( (void**)&impl->indices,
+                                  size * sizeof(CeedInt)); CeedChk(ierr);
+      impl->own_ = 1;
+
+      break;
+    case CEED_USE_POINTER:
+      impl->dindices = (CeedInt *)indices;
+      impl->indices  = NULL;
+    }
+
+  } else
+    return CeedError(r->ceed, 1, "Only MemType = HOST or DEVICE supported");
+
+  r->data    = impl;
+  r->Apply   = CeedElemRestrictionApply_Magma;
+  r->Destroy = CeedElemRestrictionDestroy_Magma;
+
+  return 0;
+}
+
+// Contracts on the middle index
+// NOTRANSPOSE: V_ajc = T_jb U_abc
+// TRANSPOSE:   V_ajc = T_bj U_abc
+// If Add != 0, "=" is replaced by "+="
+static int CeedTensorContract_Magma(Ceed ceed,
+                                    CeedInt A, CeedInt B, CeedInt C, CeedInt J,
+                                    const CeedScalar *t, CeedTransposeMode tmode,
+                                    const CeedInt Add,
+                                    const CeedScalar *u, CeedScalar *v) {
+#ifdef USE_MAGMA_BATCH
+  magma_dtensor_contract(ceed, A, B, C, J, t, tmode, Add, u, v);
+#else
+  CeedInt tstride0 = B, tstride1 = 1;
+  if (tmode == CEED_TRANSPOSE) {
+    tstride0 = 1; tstride1 = J;
+  }
+  CeedDebug("\033[31m[CeedTensorContract] A=%d, J=%d, C=%d, B=%d: %d %d %d",
+            A,J,C,B,A*J*B*C, C*J*A, C*B*A);
+  for (CeedInt a=0; a<A; a++) {
+    for (CeedInt j=0; j<J; j++) {
+      if (!Add) {
+        for (CeedInt c=0; c<C; c++)
+          v[(a*J+j)*C+c] = 0;
+      }
+      for (CeedInt b=0; b<B; b++) {
+        for (CeedInt c=0; c<C; c++) {
+          v[(a*J+j)*C+c] += t[j*tstride0 + b*tstride1] * u[(a*B+b)*C+c];
+        }
+      }
+    }
+  }
+#endif
+  return 0;
+}
+
+static int CeedBasisApply_Magma(CeedBasis basis, CeedTransposeMode tmode,
+                                CeedEvalMode emode,
+                                const CeedScalar *u, CeedScalar *v) {
+  int ierr;
+  const CeedInt dim = basis->dim;
+  const CeedInt ndof = basis->ndof;
+  const CeedInt nqpt = ndof*CeedPowInt(basis->Q1d, dim);
+  const CeedInt add = (tmode == CEED_TRANSPOSE);
+
+  CeedDebug("\033[01m[CeedBasisApply_Magma] vsize=%d",
+            ndof*CeedPowInt(basis->P1d, dim));
+
+  if (tmode == CEED_TRANSPOSE) {
+    const CeedInt vsize = ndof*CeedPowInt(basis->P1d, dim);
+    for (CeedInt i = 0; i < vsize; i++)
+      v[i] = (CeedScalar) 0;
+  }
+  if (emode & CEED_EVAL_INTERP) {
+    CeedInt P = basis->P1d, Q = basis->Q1d;
+    if (tmode == CEED_TRANSPOSE) {
+      P = basis->Q1d; Q = basis->P1d;
+    }
+    CeedInt pre = ndof*CeedPowInt(P, dim-1), post = 1;
+    CeedScalar tmp[2][ndof*Q*CeedPowInt(P>Q?P:Q, dim-1)];
+    CeedDebug("\033[01m[CeedBasisApply_Magma] tmpsize = %d",
+              ndof*Q*CeedPowInt(P>Q?P:Q, dim-1));
+    for (CeedInt d=0; d<dim; d++) {
+      ierr = CeedTensorContract_Magma(basis->ceed, pre, P, post, Q, basis->interp1d,
```

Comment:
So this is called once per element, and internally does `magma_malloc` and other relatively heavy operations for that granularity (especially if targeting a GPU). I had assumed that the MAGMA implementation would want hoist this call up a level to batch over many elements.

Reply:
Yes, definitely, this is the plan, but left this part for next release. For now we wanted to understand the flow, and have the functionality coverage on how magma will get plugged in, and have a way to fuse kernels over an element and have device code looking like the CPU code.